pkg/controller/resourcedistribution/resourcedistribution_controller.go

```diff
@@ -0,0 +1,368 @@
+func (r *ReconcileResourceDistribution) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	klog.V(3).Infof("ResourceDistribution(%s) begin to reconcile", req.NamespacedName.Name)
+	// fetch resourcedistribution instance as distributor
+	distributor := &appsv1alpha1.ResourceDistribution{}
+	if err := r.Client.Get(context.TODO(), req.NamespacedName, distributor); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	return r.doReconcile(distributor)
+}
+
+// doReconcile distribute and clean resource
+func (r *ReconcileResourceDistribution) doReconcile(distributor *appsv1alpha1.ResourceDistribution) (ctrl.Result, error) {
+	matchedNamespaces, unmatchedNamespaces, err := getNamespaceForDistributor(r.Client, &distributor.Spec.Targets)
+	if err != nil {
+		klog.Errorf("ResourceDistribution parse namespaces error, err: %v, name: %s", err, distributor.Name)
+		return reconcile.Result{}, err
+	}
+	resource, errs := utils.DeserializeResource(&distributor.Spec.Resource, field.NewPath("resource"))
+	if len(errs) != 0 || resource == nil {
+		return reconcile.Result{}, errs.ToAggregate()
+	}
+
+	// 1. distributor distribute resource to matched namespaces
+	succeeded, distributeErrs := r.distributeResource(distributor, matchedNamespaces, resource)
+
+	// 2. distributor clean its owned resources in unmatched namespaces
+	cleanErrs := r.cleanResource(distributor, unmatchedNamespaces, resource)
+
+	// 3. process all errors about resource distribution and clean
+	conditions, handleResourceError := r.handleErrors(distributor.Name, distributeErrs, cleanErrs)
+
+	// 4. update distributor status
+	newStatus := &appsv1alpha1.ResourceDistributionStatus{}
+	setNewStatus(newStatus, distributor, conditions, int32(matchedNamespaces.Len()), succeeded)
+	if !reflect.DeepEqual(distributor.Status, *newStatus) {
+		if updateStatusErr := r.updateDistributorStatus(distributor.Name, newStatus); updateStatusErr != nil {
+			return reconcile.Result{}, fmt.Errorf("handle resource error: %v, update status error: %v", err, updateStatusErr)
+		}
+	}
+	return ctrl.Result{}, handleResourceError
+}
+
+func (r *ReconcileResourceDistribution) distributeResource(distributor *appsv1alpha1.ResourceDistribution,
+	matchedNamespaces sets.String, resource runtime.Object) (int32, <-chan *UnexpectedError) {
+
+	var succeeded int32
+	wait := sync.WaitGroup{}
+	wait.Add(len(matchedNamespaces))
+	errCh := make(chan *UnexpectedError, len(matchedNamespaces))
+	newResourceHashCode := hashResource(distributor.Spec.Resource)
+	// begin to distribute and update resource
+	for namespaceName := range matchedNamespaces {
+		go func(namespace string) {
+			defer wait.Done()
+
+			resourceName := utils.ConvertToUnstructured(resource).GetName()
```

Comment:
move ConvertToUnstructured ( and even makeResourceObject )out of the namespace iteration. The resource should be almost identical for different namespace